upb/encode.c

```diff
@@ -448,11 +448,17 @@ static bool encode_shouldencode(upb_encstate* e, const upb_Message* msg,
         memcpy(&ch, mem, 1);
         return ch != 0;
       }
+#if UINTPTR_MAX == 0xffffffff
+      case upb_FieldRep_Pointer:
+#endif
       case kUpb_FieldRep_4Byte: {
         uint32_t u32;
         memcpy(&u32, mem, 4);
         return u32 != 0;
       }
+#if UINTPTR_MAX != 0xffffffff
+      case kUpb_FieldRep_Pointer:
```

Comment:
instead of using the preprocessor, how about
```
case Pointer:
  uintptr_t ptr;
  memcpy(&ptr, mem, sizeof(uintptr_t));
  return ptr != 0;
```

Reply:
My thought here was to reduce the number of branch targets and therefore improve branch predictor performance. This seems to be borne out by benchmarks: when I try this variant encoding performance degrades by ~5%.

In the longer-term, I want a faster, hasbit-driven encode that will not require this check at all.

Reply:
SGTM